Source/WebCore/page/Page.cpp

```diff
@@ -2124,11 +2124,11 @@ void Page::setDebugger(JSC::Debugger* debugger)
         frame->windowProxy().attachDebugger(m_debugger);
 }
 
-StorageNamespace* Page::sessionStorage(bool optionalCreate)
+StorageNamespace* Page::sessionStorage(SecurityOrigin& topOrigin, bool optionalCreate)
 {
-    if (!m_sessionStorage && optionalCreate) {
+    if (!m_sessionStorage || !m_sessionStorage->topLevelOrigin() || !m_sessionStorage->topLevelOrigin()->isSameOriginDomain(topOrigin)) {
         ASSERT(m_settings->sessionStorageQuota() != StorageMap::noQuota);
-        m_sessionStorage = m_storageNamespaceProvider->createSessionStorageNamespace(*this, m_settings->sessionStorageQuota());
+        m_sessionStorage = m_storageNamespaceProvider->maybeCreateSessionStorageNamespace(topOrigin, *this, m_settings->sessionStorageQuota(), optionalCreate);
```

Comment:
This does not look correct. When a SessionStorageNamespace object is destroyed, its StorageAreaMap objects will be destroyed, and StorageAreaImpl (Storage) objects that weakly points to the StorageAreaMap will be numb. That means when third party iframe accesses sessionStorage, the main frame's existing sessionStorage will be not usable. I think you can write a test to see if it's true.

I think it's better do what TransientLocalStorage does: let StorageNamespaceProvider hold strong ref to SessionStorageNamespace. 
Or if you want SessionStorageNamespace to go away when it's no in use, let StorageNamespaceProvider hold weak ref to SessionStorageNamespace, and let DOMWindow hold strong ref to SessionStorageNamespace.

Reply:
A third party iframe should have the same namespace as the main frame's namespace because the namespace is keyed on the main frame's origin and the page (session storage) identifier, so I don't believe that should be a problem.

I originally tried replicating how TransientLocalStorage works, but I stopped because I incorrectly thought the Page object and SessionNamespaceProvider object had different lifetimes. I just went back and implemented it, and now I realize this has very similar behavior as letting Page hold the strong ref to the StorageNamespace because Page and StorageNamespaceProvider have the same lifetime (Page holds the strong ref to the namespace provider). This should be the same behavior as the existing implementation in the case where Page is destroyed before DOMWindow, as well. I'm sorry if I'm misunderstanding something.

Reply:
Oh you are right that 3rd party frame won't cause SessionStorageNamespace to be destroyed! Sorry for the confusion. But I think cross-origin navigation can still cause the issue. I added a test to the bug https://bugs.webkit.org/show_bug.cgi?id=210776. Please check if it passes with this change.